### DISCUSSION TITLE
Importing in electron, ts, webpack app comes back as undefined

### DISCUSSION THREAD

**Discussion by samlhuillier** [withcatai/node-llama-cpp#81] [General]
When I try to do this in my electron app:
import { LlamaModel, LlamaContext, LlamaChatSession } from "node-llama-cpp";

console.log(LlamaModel);
console.log(LlamaContext);
console.log(LlamaChatSession);

I get back each of them being undefined. Anyone know how to fix?

**Comment by samlhuillier** [comment]
Looks like this works:
require("node-llama-cpp").then((nodeLLamaCpp: any) => {
      const model = new nodeLLamaCpp.LlamaModel({

Closing discussion :)

**Reply by v4lentin1879** [reply]
for me this results still in a undefined error. Any ideas?
```
const summary = await require("node-llama-cpp").then((module: any) => {
                const { getLlama, LlamaChatSession, LlamaGrammar } = module;
...
                return session.prompt(prompt);

        });
```

**Reply by giladgd** [reply]
@v4lentin1879 Run this command to generate an Electron project that uses `node-llama-cpp`:
```bash
npm create --yes node-llama-cpp@beta
```

**Reply by v4lentin1879** [reply]
Thank you! However, I'm still looking for a solution. I'm developing a vscode extension (electron v28) which currently does not support esm modules. Thus, I need to import the node-llama-cpp dynamically. This works fine until I package the extension. After that, I always get an error "Invalid Host defined Options". This seems to be a known issue with dynamic imports in vscode. Is there any other way I could import a esm module in cjs?

**Reply by giladgd** [reply]
@v4lentin1879 This limitation seems to be imposed by [VS Code Loder](https://github.com/microsoft/vscode-loader/issues/36) as it runs each extension in an isolated environment using [node's `vm` module](https://nodejs.org/api/vm.html) without setting [`importModuleDynamically`](https://nodejs.org/api/vm.html#support-of-dynamic-import-in-compilation-apis), which makes it impossible to load ESM modules.

Due to the nature of how `node-llama-cpp` works, the structure of the files of the module must be preserved in order for it to work properly, so it'd have to be packed into the VS Code extension in its original structure, and using a bundler to bundle all of its code into one file cannot be done as it would break it (you should mark `node-llama-cpp` as an external module in your bundler of choice).

IMO, the best way forward for this is to add support for importing ESM modules in extensions in the VS Code loader.

I've looked at the code of VS Code Loader a bit, and it seems that creating a PR for this shouldn't be too complicated.
There are [4 usages of node's `vm` Script](https://github.com/search?q=repo:microsoft/vscode-loader%20.Script&type=code) there, and each of them should be supplied with an added [`importModuleDynamically`](https://nodejs.org/api/vm.html#when-importmoduledynamically-is-a-function) in the options, to supply the logic that will actually make the dynamic import work on the Electron side, and the resolving of the module to import should be done relatively to the script file path (to make sure the requested module/file is resolved from the `node_modules` folder of the extension and not from VS Code's `node_modules` folder for example).

It's important to detect whether the current code is running in the main process or the renderer process, as using a dynamic `import` of a node module on the renderer can crash the process (from my experience with it).
Since extensions are supposed to be loaded on the main process, this shouldn't be an obstacle, but it's important to handle that case.

I would have done this myself, but I'm not sure I'll have enough time for this soon, since I'm currently working on preparing `node-llama-cpp`'s beta version for a stable release, and it's taking all of my time.

**Reply by v4lentin1879** [reply]
@giladgd Thank you so much! I will look into this and raise a pr then.. 😁

**Reply by v4lentin1879** [reply]
@giladgd Another idea: Do you think it would be possible to fork node-llama-cpp and change it to use CommonJS?

**Reply by giladgd** [reply]
@v4lentin1879 `node-llama-cpp` depends on some modules that are also ESM-only, so doing that would require rewriting a lot of the code to work with modules that support CommonJS.
Since the structure of the files is necessary for `node-llama-cpp` to function properly, bundling `node-llama-cpp` in its current form with its dependencies via CommonJS is impractical as it will break it.
Also, `node-llama-cpp` uses many features available only on ESM that cannot be replicated in a stable manner in CommonJS (such as top-level await, partial circular dependency graphs (due to TypeScript type guards), etc.).

Since the node ecosystem is moving to ESM, maintaining a new CommonJS module today is inadvisable.
Thus, I reckon it's inevitable that VS Code will have to support ESM to continue being compatible with the node ecosystem.

**Reply by v4lentin1879** [reply]
I understand. Thank you very much for explaining! :)
